plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/AdjustNodeLabels.java

```diff
@@ -0,0 +1,218 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0 
+ *******************************************************************************/
+package org.eclipse.elk.alg.layered.intermediate;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.elk.alg.layered.graph.LGraph;
+import org.eclipse.elk.alg.layered.graph.LLabel;
+import org.eclipse.elk.alg.layered.graph.LNode;
+import org.eclipse.elk.alg.layered.graph.Layer;
+import org.eclipse.elk.alg.layered.options.InternalProperties;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.alg.layered.p4nodes.NetworkSimplexPlacer;
+import org.eclipse.elk.core.alg.ILayoutProcessor;
+import org.eclipse.elk.core.math.ElkPadding;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.Direction;
+import org.eclipse.elk.core.options.MergeRange;
+import org.eclipse.elk.core.util.IElkProgressMonitor;
+import org.eclipse.elk.graph.ElkLabel;
+import org.eclipse.elk.graph.ElkNode;
+
+/**
+ * @author jep
+ *
+ */
+public final class AdjustNodeLabels implements ILayoutProcessor<LGraph> {
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see org.eclipse.elk.core.alg.ILayoutProcessor#process(java.lang.Object,
+     * org.eclipse.elk.core.util.IElkProgressMonitor)
+     */
+    @Override
+    public void process(final LGraph graph, final IElkProgressMonitor monitor) {
+        monitor.begin("Label adjustment", 0);
+
+        // combine node labels if possible
+        Direction direction = graph.getProperty(LayeredOptions.DIRECTION);
+        if (direction == Direction.DOWN || direction == Direction.UP) {
+            // node sizes and label sizes and positions are transposed here
+            down(graph);
+        } else if (direction == Direction.RIGHT || direction == Direction.LEFT) {
+            right(graph);
+        }
+
+        // recalculate size of labels and nodes
+        LabelManagementProcessor labelManager = new LabelManagementProcessor(false);
+        labelManager.process(graph, monitor);
+        LabelAndNodeSizeProcessor nodeSizeProcessor = new LabelAndNodeSizeProcessor();
+        nodeSizeProcessor.process(graph, monitor);
+        NetworkSimplexPlacer placer = new NetworkSimplexPlacer();
+        placer.process(graph, monitor);
+
+        monitor.done();
+    }
+
+    /**
+     * Combines labels if the node size allows it.
+     * 
+     * @param graph
+     *            The graph.
+     */
+    private void down(final LGraph graph) {
```

Comment:
I don't think having two methods 'down' and 'right' here is too sensible. Either one common method that can be transposed later or case checks within the method would make sense, since they are essentially doing the same thing in different directions.